test/infra/proxy.py

```diff
@@ -0,0 +1,98 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+import asyncio
+from typing import Tuple
+
+import aiotools
+from loguru import logger as LOG
+
+from .async_utils import EventLoopThread, race_tasks
+
+
+class Proxy(EventLoopThread):
+    """
+    A TCP proxy, running as a context manager.
+
+    The proxy listens for connections on a socket, and for each incoming
+    connection creates a new outgoing connection to a pre-configured upstream
+    host and port number. It then forwards data back and forth between the two
+    sockets.
+
+    The configured upstream can be modified after instantiation by calling
+    `set_upstream`. The change only affects future connections, not existing
```

Comment:
If and when the upstream shuts down, it closes all its connections, and in-turn the proxy shuts down the corresponding incoming connection. So this works because in our use-case where we shut down the old service.

If we were to use this to switch between running services we might run into the issue you've described. I'm not sure it is worth fixing this now, although I'll add a note to explain this.